Disk read resource

SirTyson opened the pull request:
# Description

Adds support for p23 disk read resource in `checkValid`. Also changes `Resource` to use disk read entries in p23.

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v8.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [ ] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

dmkozh commented on src/transactions/TransactionFrame.cpp:
```diff
@@ -59,6 +59,35 @@ namespace
+{
+    // First count classic entry reads
+    uint32_t count = 0;
+    auto countClassic = [&count](auto keys) {
```
Should this be `auto const&`? I'm not sure if this might end up copying the footprint...

dmkozh commented on src/transactions/TransactionFrame.cpp:
```diff
@@ -601,13 +641,36 @@ TransactionFrame::checkSorobanResources(
+        {
+            diagnosticEvents.pushError(
+                SCE_STORAGE, SCEC_EXCEEDED_LIMIT,
+                "transaction total-entry-read resources exceed network config "
```
nit: Since this has been renamed to maxFootprintEntries, the error message should stay the same ('the number of entries in transaction footprint exceeds the network config limit')

dmkozh commented on src/transactions/TransactionFrame.cpp:
```diff
@@ -601,13 +641,36 @@ TransactionFrame::checkSorobanResources(
             SCE_STORAGE, SCEC_EXCEEDED_LIMIT,
-            "transaction entry-read resources exceed network config limit",
-            {makeU64SCVal(readEntries.size() + writeEntries.size()),
+            "transaction entry-disk-read resources exceed network config limit",
```
I think we should not change this error message for p22; it's ok to just have a protocol condition here and copy-paste the error text with the slight adjustment.

dmkozh commented on src/herder/TransactionQueue.h:
```diff
@@ -139,23 +139,24 @@ class TransactionQueue
 #ifdef BUILD_TESTS
     AddResult tryAdd(TransactionFrameBasePtr tx, bool submittedFromSelf,
-                     bool isLoadgenTx = false);
+                     uint32_t ledgerVersion, bool isLoadgenTx = false);
```
It's kind of annoying that we need to pass the protocol version to every method. I think we can just store `ledgerVersion` in tx queue and update it when we handle a processed ledger (we do a bunch of upkeep every ledger anyways).

dmkozh replied:
FWIW this could be done as a followup; it would reduce the number of diffs quite a bit though.

marta-lokhova replied:
+1 I don't think we need to pipe ledgerVersion to TransactionQueue or store it as a member var -- just use mApp.getLedgerManager().getLastClosedLedgerHeader().header.ledgerVersion (there are already places in TransactionQueue that do that, e.g. in canAdd)

marta-lokhova commented on src/herder/TxQueueLimiter.h:
```diff
@@ -57,7 +59,8 @@ class TxQueueLimiter
         std::vector<std::pair<TransactionFrameBasePtr, bool>> const& txsToEvict,
         TransactionFrameBase const& txToFit,
-        std::function<void(TransactionFrameBasePtr const&)> evict);
+        std::function<void(TransactionFrameBasePtr const&)> evict,
```
Same thing here: I don't think you need to pipe ledgerVersion.

marta-lokhova commented on src/herder/test/HerderTests.cpp:
```diff
@@ -3190,6 +3190,9 @@ TEST_CASE("soroban txs each parameter surge priced", "[soroban][herder]")
     }
+    /*
+    // Loadgen currently doesn't support generating soroban load with disk based
+    resources in p23
```
ideally, at the time of the upgrade, the tx queue can be full, and new p23 transactions come in so we can verify the footprint semantics change works as expected

marta-lokhova commented on src/herder/TxQueueLimiter.h:
```diff
@@ -40,8 +40,10 @@ class TxQueueLimiter
-    void removeTransaction(TransactionFrameBasePtr const& tx);
+    void addTransaction(TransactionFrameBasePtr const& tx,
+                        uint32_t ledgerVersion);
+    void removeTransaction(TransactionFrameBasePtr const& tx,
```
It looks like we're retroactively changing the semantics of tx queue. That is, transactions added in p22 followed p22 semantics for validity, and limits. In p23, these rules are changing. For validity, this shouldn't be a problem since invalid transactions will just get purged from the queue (what we discussed yesterday). For limits though, we might be violating tx queue/limiter invariance with p23. It's hard to tell if this will work correctly, since we don't have tests for this yet (I think we should). We were also discussing with @dmkozh that at protocol upgrade it's probably safer to just rebuild the tx queue with the new limit semantics. It will cost a little to rebuild the queue, but at least we can be confident we won't trip some unexpected invariant.

SirTyson replied:
Wrt to the 23 upgrade, limits are strictly getting laxer, so there is no limit upper bound that we could invalidate on p22 -> 23, I'm not sure if there are some sort of checks on lower bound or expected limit capacity though. In general though I agree it's probably a good idea to rebuild the queue on updates. Is this not an issue for regular config upgrades as well that has existed for a while?

marta-lokhova replied:
yeah I think your reasoning makes sense. Conceptually it should work. We just don't have enough info/tests to be confident it will work and won't trip anything. So we need to increase confidence. What we could do is (from discussion with @dmkozh):
* Rebuild the queue post p23 upgrade (note that this has to be done synchronously to avoid new transactions sliding in before the rebuild). You could reset the tx queue instance and re-add all existing transactions, and surge pricing logic will take care of sorting/evictions. 
* Write a test that submits load before, during and after the upgrade, and verify nothing crashes and all transactions get into the ledger. Note that loadgen should be configured such that we always have a full or nearly full tx queue at the time of the upgrade.
* Since this is time sensitive, in RC1 we _could_ just drop the whole queue at the time of p23 upgrade without re-building. This is fine for testing and places like futurenet. For testnet and pubnet, we'd want to do the right thing and rebuild the queue. 

Wdyt?

marta-lokhova replied:
For the existing network upgrades, this seems just as footgun-y. I can't tell if it actually works, or if it worked historically because we've always upgraded limits in one direction (increasing), or only touched a subset of limits that didn't trip anything. Either way, it seems safer to just rebuild the queue, so we can at least reason about what would happen.

SirTyson commented:
I've pushed fixes to address y'alls comments. Wrt to dropping the tx queue, this should basically be equivalent to increasing limits via a network config. While it's not introducing anything new, limit upgrades is a blind spot in our testing/might be buggy if limits aren't strictly increasing. I guess course of action is to reset the queue in this PR for rc1 then add better end-to-end tests for upgrade/limit changes for rc2?

dmkozh commented:
> I guess course of action is to reset the queue in this PR for rc1 then add better end-to-end tests for upgrade/limit changes for rc2?

I've looked a bit at the supercluster tests, and we do a bunch of tests with protocol upgrades, and if we just do a 'quick hack' by resetting the queue, we risk breaking these. We also have https://github.com/stellar/supercluster/blob/8412dffb7ac37422221dca988252e68286d2d0ad/src/FSLibrary/MissionSorobanConfigUpgrades.fs#L48 mission that tests that the network operates fine in case if settings are increased. Given that this change doesn't reduce the tx queue capacity and that we know that not decreasing capacity is safe, I'd say that we have two options:
- Keep PR as is and rely on the tx queue flexibility. There is a slight risk of some tx queue invariant breaking, but we should find out with supercluster
- Clear and rebuild the tx queue on protocol upgrade. I think this is a safer option in general, but I'm not sure it's mandatory for RC1.

marta-lokhova commented:
it seems fine to do the proper rebuild in RC2 while keeping things as-is in RC1 in order to unstick this change. We can also trigger a supercluster run already to get some early integration tests signals. Looks like GH actions are currently failing though.

marta-lokhova commented:
please note, that an end-to-end p22->p23 test with a full tx queue is still needed, but that can be added in RC2 as well.

SirTyson commented on src/ledger/NetworkConfig.h:
```diff
@@ -152,7 +152,8 @@ struct InitialSorobanNetworkConfig
 
     // Ledger cost extension settings
-    static constexpr uint32_t TX_MAX_IN_MEMORY_READ_ENTRIES = 100;
+    static constexpr uint32_t TX_MAX_IN_MEMORY_READ_ENTRIES =
```
I ended up cherry-picking your PR for this. Reasoning being, some tests don't perform a protocol upgrade and assume the default starting limits. I could change all these tests, but cherry-picking was easier and we'll need to do it anyway.

SirTyson commented:
I've updated this branch with the release XDR and env hashes and fixed a couple test issues around network configs. I'm starting an SSC run to make sure TX queue still works on upgrade (at least well enough for rc1) and opened a follow up issue for rc2 https://github.com/stellar/stellar-core/issues/4740.
